utilities/transactions/write_unprepared_txn_db.cc

```diff
@@ -367,29 +365,49 @@ Iterator* WriteUnpreparedTxnDB::NewIterator(const ReadOptions& options,
   // TODO(lth): Refactor so that this logic is shared with WritePrepared.
   constexpr bool ALLOW_BLOB = true;
   constexpr bool ALLOW_REFRESH = true;
-  std::shared_ptr<ManagedSnapshot> own_snapshot = nullptr;
   SequenceNumber snapshot_seq;
   SequenceNumber min_uncommitted = 0;
-  if (options.snapshot != nullptr) {
-    snapshot_seq = options.snapshot->GetSequenceNumber();
-    min_uncommitted =
-        static_cast_with_check<const SnapshotImpl, const Snapshot>(
-            options.snapshot)
-            ->min_uncommitted_;
-  } else {
-    auto* snapshot = GetSnapshot();
-    // We take a snapshot to make sure that the related data in the commit map
-    // are not deleted.
-    snapshot_seq = snapshot->GetSequenceNumber();
-    min_uncommitted =
-        static_cast_with_check<const SnapshotImpl, const Snapshot>(snapshot)
-            ->min_uncommitted_;
-    own_snapshot = std::make_shared<ManagedSnapshot>(db_impl_, snapshot);
+  // TODO(lth): Currently, we can only create iterators with the transaction
+  // snapshot. This is because the iterator logic relies on snapshot validation
+  // to ensure that no committed keys exist between snapshot_seq < seq <
+  // max_visible_seq. This assumption is currently upheld by snapshot
+  // validation, and validation is run against the transaction snapshot sequence
+  // number.
+  //
+  // Therefore, we cannot guarantee correct results for iterators with no
+  // transaction snapshot, but with tracked keys. The second condition is so
+  // that we can avoid silently falling back to transaction snapshot when users
+  // have specified a snapshot in ReadOptions that we can't use. It's probably
+  // more user friendly to just fail.
+  //
+  // It may be possible to modify the iterator logic so that it does not rely on
+  // that assumption. The problem is that during Prev(), we are simply looking
+  // for, in ascending seqno order, the first non-visible key and stopping after
+  // that. However, an unprep key could be at snapshot_seq < non_visible_seq <
+  // unprep_seq, so we never reach the unprep_seq.
+  //
+  auto txn_snapshot = txn->GetSnapshot();
+  if ((txn_snapshot == nullptr && txn->GetNumKeys() > 0) ||
```

Comment:
Multi-line expressions are very hard to understand. Is it possible to split it to multiple conditions, each with its own boolean variable with a self-describing name?